content/docs/using-pulumi/debugging/_index.md

```diff
@@ -0,0 +1,299 @@
+---
+title_tag: "Debugging Pulumi Programs"
+meta_desc: "Guide to attaching a debugger to Pulumi programs and stepping through the code."
+title: Debugging
+h1: Debugging Pulumi programs
+meta_image: /images/docs/meta-images/docs-meta.png
+menu:
+    usingpulumi:
+        identifier: debugging
+        weight: 9
+---
+
+Because Pulumi uses general purpose programming languages to provision cloud resources, you can take advantage of native debugging tools to troubleshoot your infrastructure definitions.
+
+Note that you can't directly F5-launch your Pulumi program from your IDE. Instead, you run the Pulumi CLI that will launch your program via its language runtime in a separate process, and then you attach to that process from your IDE. See [How Pulumi works](/docs/concepts/how-pulumi-works/) to learn more about Pulumi execution mode.
+
+The tools differ depending on the language you are using, but the process stays similar:
+
+1. Start with an existing Pulumi project that contains `Pulumi.yaml` and your program.
+2. For some languages, you'll have to temporarily modify your program to pause until a debugger is attached.
+3. Open your IDE and set a breakpoint in your program.
+4. Run `pulumi up` from a command line. Note that `pulumi up` runs your program twice: once for preview, and once for update, and both will trigger the debugger. Use `pulumi preview` or `pulumi up --skip-preview` to pick one of the two modes.
+5. Wait until `pulumi` is paused waiting for a debugger to attach.
+6. Attach the debugger to a running process of your language runtime.
+7. Step through the program to find the problem.
+8. Run the program to completion and let `pulumi` shut down all the processes.
+9. Change your program and repeat the process if needed.
+
+Subsequent sections outline the specifics for each language runtime.
+
+## Debugging Node.js programs
+
+Run the Pulumi CLI by passing the [`--inspect-brk`](https://nodejs.org/en/learn/getting-started/debugging#command-line-options) option to node:
+
+```bash
+NODE_OPTIONS='--inspect-brk' pulumi up
+```
+
+The stack execution will pause until a debugger is attached:
+
+```plain
+Previewing update (dev)
+
+     Type                      Name           Plan       Info
+ +   pulumi:pulumi:Stack       aws-ts-dev     create     For help, see: https://nodejs.org/en/docs/inspector
+```
+
+Open your Pulumi program in your IDE and set a breakpoint at the line where you want the debugger to stop.
+
+Then, attach your debugger to the process with the `node` runtime. For example, in [VS Code](https://code.visualstudio.com/docs/editor/debugging), you can define the following `.vscode/launch.json` file:
```

Comment:
Should we recommend using the JavaScript Debug Terminal if using VSCode - there's zero additional config required.

Reply:
Added a mention of auto-attach and debug terminal